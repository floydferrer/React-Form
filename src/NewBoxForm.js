import React, { useState } from "react";

const NewBoxForm = ({ addBox }) => {
    const initialFormState = {
        color: '',
        width: '',
        height: ''
    }
    const initialInvalidState = {
        color: true,
        width: true,
        height: true
    }
    const initialTouchedState = {
        color: false,
        width: false,
        height: false
    }
    const [formData, setFormData] = useState(initialFormState);
    const [isInvalid, setIsInvalid] = useState(initialInvalidState);
    const [isTouched, setIsTouched] = useState(initialTouchedState);
    const handleChange = e => {
        setIsTouched(prev => ({
            ...prev,
            [e.target.name]: true
        }));
        // setIsTouched(isTouched[e.target.name] = true);
        const {name, value} = e.target;
        if(value === '') {
            setIsInvalid(prev => ({
                ...prev,
                [e.target.name]: true
            }));
        } else {
            setIsInvalid(prev => ({
                ...prev,
                [e.target.name]: false
            }));
        }

        setFormData(data => ({
            ...data,
            [name]: value
        }))
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        const { color } = formData;
        if(!isInvalid.color && !isInvalid.width && !isInvalid.height) {
            // alert(`${color} box added!`);
            addBox({...formData})
            setFormData(initialFormState);
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <h3>Create Box</h3>
            <label htmlFor="color">Color</label>
            <input
                type="text"
                placeholder="color"
                name="color"
                id="color"
                value={formData.color}
                onChange={handleChange}
            />
            {isInvalid.color && isTouched.color && <span style={{color: 'red'}}>Color cannot be blank!</span>}
            
            <label htmlFor="width">Width</label>
            <input
                type="number"
                placeholder="width"
                name="width"
                id="width"
                value={formData.width}
                onChange={handleChange}
            />
            {isInvalid.width && isTouched.width && <span style={{color: 'red'}}>Width cannot be blank!</span>}
            
            <label htmlFor="height">Height</label>
            <input
                type="number"
                placeholder="height"
                name="height"
                id="height"
                value={formData.height}
                onChange={handleChange}
            />
            {isInvalid.height && isTouched.height && <span style={{color: 'red'}}>Height cannot be blank!</span>}
            <button role="button">Add Box</button>
        </form>
    )
}

export default NewBoxForm;