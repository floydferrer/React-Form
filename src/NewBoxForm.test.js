import { render, fireEvent } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm";

it("renders without crashing", function() {
    render(<NewBoxForm />)
});

it("matches snapshot", function() {
    const { asFragment } = render(<NewBoxForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("works when you submit form", function() {
    const addBoxMock = jest.fn();
    const { getByLabelText, getByRole } = render(<NewBoxForm addBox={addBoxMock}/>)
    const colorInput = getByLabelText("Color");
    fireEvent.change(colorInput, { target: { value: 'test-yellow' }});
    const widthInput = getByLabelText("Width");
    fireEvent.change(widthInput, { target: { value: '200' }});
    const heightInput = getByLabelText("Height");
    fireEvent.change(heightInput, { target: { value: '200' }});
    const addBtn = getByRole('button');
    fireEvent.click(addBtn);

    expect(addBoxMock).toHaveBeenCalled();
  });