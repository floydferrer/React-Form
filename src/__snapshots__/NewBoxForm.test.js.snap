// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <h3>
      Create Box
    </h3>
    <label
      for="color"
    >
      Color
    </label>
    <input
      id="color"
      name="color"
      placeholder="color"
      type="text"
      value=""
    />
    <label
      for="width"
    >
      Width
    </label>
    <input
      id="width"
      name="width"
      placeholder="width"
      type="number"
      value=""
    />
    <label
      for="height"
    >
      Height
    </label>
    <input
      id="height"
      name="height"
      placeholder="height"
      type="number"
      value=""
    />
    <button
      role="button"
    >
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
